src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java

```diff
@@ -235,10 +234,11 @@ public QuerySearchResultProvider load(final ShardSearchRequest request, final Se
             }
         } else {
             key.shard.queryCache().onHit();
+            // restore the cached query result into the context
+            final QuerySearchResult result = context.queryResult();
+            result.readFromWithId(context.id(), value.streamInput());
```

Comment:
I had to deserialize the shard response all the time on the shard itself in order to figure out how many hits the shard had (to know if I can close the context). So I removed BytesQuerySearchResult which was not helping anymore.

Another idea would be to store the number of hits on the side of BytesQuerySearchResult but I'm concerned the benefit/complexit ratio would not be good?